### PR TITLE
daemon: Daemon should handle npm packages from different sources

### PR DESCRIPTION
**Pull request by rickycao-qy**
Refer to #579 

Todo: unit tests

### CODE REVIEW

File: packages/costa/package.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
     "better-opn": "^1.0.0",
     "debug": "^4.1.1",
     "fs-extra": "^8.1.0",
+    "md5": "^2.3.0",
```
Comment (yorkie) [inline packages/costa/package.json:26]: The `md5` depends on the [crypt](https://github.com/pvorb/node-crypt/blob/master/crypt.js) which is implemented in JavaScript for browser, I'd prefer do use the Node.js crypto module, see https://nodejs.org/dist/latest-v12.x/docs/api/crypto.html#crypto_crypto_createhash_algorithm_options, it supports the algorithm `md5`, too. And it should be more stable and performant.

File: packages/costa/src/runtime.ts [hunk 2]
```diff
@@ -67,13 +68,13 @@ interface InstallOptions {
...
 
-function spawnAsync(command: string, args: string[], opts: SpawnOptions, stdio: LogStdio): Promise<void> {
+function spawnAsync(command: string, args: string[], opts: SpawnOptions, stdio?: LogStdio): Promise<void> {
   return new Promise((resolve, reject) => {
     opts.stdio = [ null, 'pipe', 'pipe' ];
     opts.detached = false;
     const child = spawn(command, args, opts);
-    pipeLog(child.stdout, stdio.stdout, stdio.prefix);
-    pipeLog(child.stderr, stdio.stderr, stdio.prefix);
+    stdio?.stdout && pipeLog(child.stdout, stdio.stdout, stdio.prefix);
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:76]: I think we could just verify it in the function `pipeLog()`.

**Review by yorkie — CHANGES_REQUESTED** [review]
Current there are too much flatten functions, we should refactor it to represent the flow in https://github.com/alibaba/pipcook/issues/579 in better way.

File: packages/costa/src/runtime.ts [hunk 3]
```diff
@@ -128,6 +135,22 @@ async function fetchPackageJsonFromTarballStream(fileStream: NodeJS.ReadableStre
   return JSON.parse(packageJson);
 }
 
+async function fetchMd5(filePath: string): Promise<string> {
+  return new Promise((resolve, reject) => {
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:139]: Is there necessary to use the stream API? It seems that the file might be a small package.json file thus digest() looks simple?

File: packages/costa/package.json [hunk 4]
```diff
@@ -31,6 +31,7 @@
     "@types/debug": "^4.1.5",
     "@types/fs-extra": "^8.0.1",
     "@types/jasmine": "^3.5.7",
+    "@types/md5": "^2.2.0",
```
Comment (yorkie) [inline packages/costa/package.json:34]: This should be removed, too :)

File: packages/costa/src/runtime.ts [hunk 5]
```diff
@@ -552,6 +584,11 @@ export class CostaRuntime {
         // TODO(feely): not implement the version
         installDir, `node_modules/${pkg.name}@${pkg.version}`)
     };
+    if (source && source.from !== 'npm') {
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:587]: Source check and assignments should be moved forward the fetch* methods.

File: packages/costa/src/utils.ts [hunk 6]
```diff
@@ -12,6 +12,9 @@ export interface LogStdio {
  * @param prefix the log prefix
  */
 export function pipeLog(readable: NodeJS.ReadableStream, writable: NodeJS.WritableStream, prefix?: string): void {
```
Comment (yorkie) [inline packages/costa/src/utils.ts:14]:
```suggestion
export function pipeLog(readable?: NodeJS.ReadableStream, writable?: NodeJS.WritableStream, prefix?: string): void {
```

File: packages/costa/src/runtime.ts [hunk 7]
```diff
@@ -14,9 +14,10 @@ import {
   readJson,
   writeJson
 } from 'fs-extra';
-import { download, constants, generateId, parsePluginName } from '@pipcook/pipcook-core';
+import { download, constants, generateId, parsePluginName, PluginStatus } from '@pipcook/pipcook-core';
 import tar from 'tar-stream';
 import { spawn, SpawnOptions } from 'child_process';
+import crypto from 'crypto';
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:20]: The builtin imports should be moved behind the third-party.

File: packages/costa/src/index.ts [hunk 8]
```diff
@@ -105,6 +105,10 @@ export interface PluginPackage {
      * The plugin runtime
      */
     runtime?: 'nodejs' | 'python';
+    /**
+     * md5 of the tarball
+     */
+    md5?: string;
```
Comment (yorkie) [inline packages/costa/src/index.ts:111]: Just not sure if this should be here. /cc @FeelyChau

File: packages/costa/src/runtime.ts [hunk 9]
```diff
@@ -94,14 +95,20 @@ async function extractPackageJsonFromReadable(readable: NodeJS.ReadableStream, p
   return JSON.parse(packageJson);
 }
 
-function fetchPackageJsonFromGit(remote: string, head: string): Promise<any> {
+async function fetchPackageJsonFromGit(remote: string, head: string, source: PluginSource): Promise<any> {
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:98]: @FeelyChau @rickycao-qy Shall we use the template for these methods?

```js
const packageJson = new Package<Git | Npm | File | ...>(...);
packageJson.fetch(...);
```

File: packages/costa/src/runtime.ts (hunk 3 above)
Comment (rickycao-qy) [reply packages/costa/src/runtime.ts:139]: the file is the whole npm package and might be quite large?

File: packages/costa/package.json (hunk 4 above)
Comment (rickycao-qy) [reply packages/costa/package.json]: done

File: packages/costa/src/runtime.ts (hunk 2 above)
Comment (rickycao-qy) [reply packages/costa/src/runtime.ts]: done

File: packages/costa/src/runtime.ts (hunk 5 above)
Comment (rickycao-qy) [reply packages/costa/src/runtime.ts]: done

File: packages/costa/src/utils.ts (hunk 6 above)
Comment (rickycao-qy) [reply packages/costa/src/utils.ts]: done

File: packages/costa/src/runtime.ts (hunk 7 above)
Comment (rickycao-qy) [reply packages/costa/src/runtime.ts]: done

File: packages/costa/package.json (hunk 1 above)
Comment (rickycao-qy) [reply packages/costa/package.json]: done

**Comment by rickycao-qy** [comment]
This PR is hold for the reason that we would better to use some template methods for those functions to fetch from different sources. This to to be done

**Comment by rickycao-qy** [comment]
Pipcook will change to v2.0 with different handling way of npm packages. Will close this PR temporarily. I will open it later if necessary

**Pull request closed without merging**